File: 2017-rocks-q1/Rocks/Rocks.Fsharp.Tests/ScheduleAdherenceTests.fs
namespace Rocks.Tests

open Akka.Configuration
open Akka.FSharp
open Foq
open Akka.TestKit.Xunit2
open Xunit
open FsUnit.Xunit
open Rocks
open Rocks.Fsharp
open Rocks.Fsharp.ScheduleAdherence
open FSharp.Data.UnitSystems.SI.UnitSymbols

module ScheduleAdherenceTests =
    [<Fact>]
    let ``Sum distances with partial route completion``() =
        let route:Route = {
            RouteId = 1
            RouteDistance = 100.0<m>
            Stops = 
            [{
                Stop.StopId = 1
                DistanceAlongRoute = 10.0<m>
            };{
                StopId = 2
                DistanceAlongRoute = 20.0<m>
            }]
        }

        let stopTimes = 
            [{
                StopId = 1
                RouteId = route.RouteId
                PassengerOn = 0
                PassengerOff = 0
            }]

        SumDistances route stopTimes
        |> should equal 10.0

    [<Fact>]
    let ``Actor should emit tripexecution on unassignment``() = 
        let deps:Dependencies = {
            TripClient = Mock<ITripClient>().Create()
            RouteClient = Mock<IRouteClient>().Create()
        }

        let tripId = 1
        let routeId = 2
        let driverId = 3
        let vehicleId = 4

        let tripData:TripData = {
            PassengerCount = 0
            Trip = Some { TripId = tripId; RouteId = routeId }
            Route = 
                { 
                    RouteId = routeId
                    RouteDistance = 100.0<m>
                    Stops = 
                    [{
                        Stop.StopId = 1
                        DistanceAlongRoute = 10.0<m>
                    };{
                        StopId = 2
                        DistanceAlongRoute = 20.0<m>
                    }]
                }
            StopTimes = 
            [{
                StopId = 2
                RouteId = routeId
                PassengerOn = 0
                PassengerOff = 0
            };{
                StopId = 1
                RouteId = routeId
                PassengerOn = 0
                PassengerOff = 0
            }]
        }

        let unassign = Messages.VehicleRouteStatus {
            DriverId = driverId
            VehicleId = vehicleId
            RouteId = Some routeId
            TripId = Some tripId
            State = VehicleRouteState.NoRoute
        }

        use testkit = new TestKit()
        testkit.Sys.EventStream.Subscribe(testkit.TestActor, typeof<TripExecution>) |> ignore

        let actor = 
            spawn testkit.Sys "TestActor"
                (actorOf2Async <| States.CollectingTripData tripData deps)

        actor <! unassign
        
        let execution = testkit.ExpectMsg<TripExecution>()

        execution.TotalDistanceMeters
        |> should equal 20.0

File: 2017-rocks-q1/Rocks/Rocks.Csharp.Tests/ScheduleAdherence/ScheduleAdherenceTests.cs
using Akka.Actor;
using Akka.TestKit.Xunit2;
using Microsoft.FSharp.Collections;
using Moq;
using Rocks.Csharp.ScheduleAdherence;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Xunit;
using FluentAssertions;
using ScheduleAdherenceActor = Rocks.Csharp.ScheduleAdherence.ScheduleAdherenceActor1;
//using ScheduleAdherenceActor = Rocks.Fsharp.ScheduleAdherence.Actor;

namespace Rocks.Csharp.Tests.ScheduleAdherence
{
    public class ScheduleAdherenceTests : TestKit
    {
        [Fact]
        public void Should_emit_tripexecution_on_unassignment()
        {
            int tripId = 1;
            int routeId = 2;
            int driverId = 3;
            int vehicleId = 4;

            var trip = new Trip(tripId, routeId);
            IEnumerable<Stop> stops = new List<Stop>()
            {
                new Stop(stopId:1, distanceAlongRoute: 10),
                new Stop(stopId:2, distanceAlongRoute: 20),
            };
            var route = new Route(routeId, ListModule.OfSeq(stops), routeDistance: 100);

            var tripClientMock = new Mock<ITripClient>();
            tripClientMock.Setup(c => c.GetTrip(tripId)).ReturnsAsync(trip);

            var routeClientMock = new Mock<IRouteClient>();
            routeClientMock.Setup(c => c.GetRoute(routeId)).ReturnsAsync(route);

            var assignment = 
                new VehicleRouteStatus(
                    vehicleId, 
                    driverId, 
                    routeId.Some(), 
                    tripId.Some(), 
                    VehicleRouteState.OnRoute);

            var stopTimes =
                stops
                    .Select(s => new StopTime(s.StopId, routeId, passengerOn:0, passengerOff:0))
                    .ToList();

            var unassignment = 
                new VehicleRouteStatus(
                    vehicleId, 
                    driverId, 
                    routeId.Some(), 
                    tripId.Some(), 
                    VehicleRouteState.NoRoute);

            var actor = 
                this.ActorOf(
                    Props.Create(() =>
                        new ScheduleAdherenceActor(
                                tripClientMock.Object,
                                routeClientMock.Object)));

            this.Sys.EventStream.Subscribe(this.TestActor, typeof(TripExecution));

            actor.Tell(assignment);

            foreach (var stopTime in stopTimes)
                actor.Tell(stopTime);

            actor.Tell(unassignment);

            var execution = ExpectMsg<TripExecution>();
            execution.TotalDistanceMeters
                .Should().Be(20);
        }
    }
}
